File: app/seed.py
"""YAML-to-database seeder for cameras.

Reads ``cameras.yaml`` at startup and inserts camera records when the
database is empty, providing first-run bootstrap data.
"""

from pathlib import Path

import yaml
from loguru import logger

from app.core.config import settings
from app.core.database import session_factory
from app.core.unit_of_work import UnitOfWork
from app.domain.models import Camera


async def seed_from_yaml(yaml_path: Path) -> bool:
    """Seed the database with cameras defined in a YAML file.

    Skips seeding when the file is missing, the database already contains
    cameras, or the YAML defines none.

    Args:
        yaml_path: Path to the YAML file containing a ``cameras`` list.

    Returns:
        ``True`` when cameras were seeded, ``False`` otherwise.
    """
    if not yaml_path.exists():
        logger.warning(f"cameras.yaml not found at {yaml_path}, skipping seed")
        return False

    async with UnitOfWork(session_factory) as uow:
        existing = await uow.cameras.get_all()
        existing_by_name = {c.name: c for c in existing}

        with open(yaml_path) as f:
            data = yaml.safe_load(f)

        raw_cameras = data.get("cameras", [])
        if not raw_cameras:
            logger.warning(f"No cameras defined in YAML at {yaml_path}")
            return False

        yaml_names = {raw["name"] for raw in raw_cameras}
        for raw in raw_cameras:
            name = raw["name"]
            values = {
                "host": raw["host"],
                "port": raw.get("port", 80),
                "username": raw.get("username", ""),
                "password": raw.get("password", ""),
                "profile_token": raw.get("profile_token"),
                "snapshot_url": raw.get("snapshot_url"),
                "interval_seconds": raw.get("interval_seconds", settings.default_interval_seconds),
                "enabled": raw.get("enabled", True),
            }
            if name in existing_by_name:
                updated = await uow.cameras.update(existing_by_name[name].id, values)
                if updated:
                    logger.info(f"Updated camera: {updated.name} ({updated.host})")
            else:
                camera = Camera(name=name, **values)
                await uow.cameras.add(camera)
                logger.info(f"Seeded camera: {camera.name} ({camera.host})")

        for name, cam in existing_by_name.items():
            if name not in yaml_names:
                await uow.cameras.delete(cam.id)
                logger.info(f"Removed camera not in YAML: {name}")

    logger.info(f"Seed complete: {len(raw_cameras)} cameras synced from YAML")
    return True

File: internal/seed/seed.go
// Package seed syncs camera metadata from cameras.yaml into the
// database at startup (idempotent upsert-by-name, deleting cameras that
// were removed from the YAML). Parity with the legacy seed.py.
package seed

import (
	"context"
	"fmt"
	"os"

	"github.com/jmoiron/sqlx"
	"github.com/rs/zerolog/log"
	"gopkg.in/yaml.v3"

	"github.com/Lmex89/home-cameras/internal/domain"
	"github.com/Lmex89/home-cameras/internal/repository"
)

// yamlCamera mirrors one entry of the cameras.yaml "cameras" list.
type yamlCamera struct {
	Name            string  `yaml:"name"`
	CameraType      string  `yaml:"camera_type"`
	Host            string  `yaml:"host"`
	Port            int     `yaml:"port"`
	Username        string  `yaml:"username"`
	Password        string  `yaml:"password"`
	ProfileToken    *string `yaml:"profile_token"`
	SnapshotURL     *string `yaml:"snapshot_url"`
	DevicePath      *string `yaml:"device_path"`
	IntervalSeconds int     `yaml:"interval_seconds"`
	Enabled         bool    `yaml:"enabled"`
}

// yamlFile is the root document of cameras.yaml.
type yamlFile struct {
	Cameras []yamlCamera `yaml:"cameras"`
}

// FromYAML loads cameras from the YAML file at yamlPath and syncs them
// with the database: cameras present in both are updated, new cameras
// are inserted, and cameras missing from the YAML are deleted.
//
// Args:
//
//	ctx: Request context.
//	db: Database pool used for the sync transaction.
//	yamlPath: Absolute path of cameras.yaml.
//	defaultInterval: Fallback interval when the YAML omits it.
//
// Returns:
//
//	True when cameras were seeded/updated, false when the file is
//	missing or empty.
func FromYAML(ctx context.Context, db *sqlx.DB, yamlPath string, defaultInterval int) (bool, error) {
	data, err := os.ReadFile(yamlPath)
	if err != nil {
		if os.IsNotExist(err) {
			log.Warn().Str("path", yamlPath).Msg("cameras.yaml not found, skipping seed")
			return false, nil
		}
		return false, err
	}
	var doc yamlFile
	if err := yaml.Unmarshal(data, &doc); err != nil {
		return false, fmt.Errorf("parse cameras.yaml: %w", err)
	}
	if len(doc.Cameras) == 0 {
		log.Warn().Str("path", yamlPath).Msg("no cameras defined in YAML")
		return false, nil
	}

	tx, err := db.BeginTxx(ctx, nil)
	if err != nil {
		return false, err
	}
	defer tx.Rollback()
	repo := repository.NewCameraRepository(tx)

	existing, err := repo.GetAll(ctx)
	if err != nil {
		return false, err
	}
	existingByName := map[string]domain.Camera{}
	for _, c := range existing {
		existingByName[c.Name] = c
	}

	seeded := 0
	for _, raw := range doc.Cameras {
		interval := raw.IntervalSeconds
		if interval == 0 {
			interval = defaultInterval
		}
		cameraType := raw.CameraType
		if cameraType == "" {
			cameraType = "ip"
		}
		values := map[string]any{
			"camera_type":      cameraType,
			"host":             raw.Host,
			"port":             raw.Port,
			"username":         raw.Username,
			"password":         raw.Password,
			"profile_token":    raw.ProfileToken,
			"snapshot_url":     raw.SnapshotURL,
			"device_path":      raw.DevicePath,
			"interval_seconds": interval,
			"enabled":          raw.Enabled,
		}
		if cam, ok := existingByName[raw.Name]; ok {
			if _, err := repo.Update(ctx, cam.ID, values); err != nil {
				return false, err
			}
			log.Info().Str("camera", raw.Name).Msg("seed: updated camera")
		} else {
			newCam := &domain.Camera{
				Name:            raw.Name,
				CameraType:      cameraType,
				Host:            raw.Host,
				Port:            raw.Port,
				Username:        raw.Username,
				Password:        raw.Password,
				ProfileToken:    raw.ProfileToken,
				SnapshotURL:     raw.SnapshotURL,
				DevicePath:      raw.DevicePath,
				IntervalSeconds: interval,
				Enabled:         raw.Enabled,
			}
			if err := repo.Add(ctx, newCam); err != nil {
				return false, err
			}
			log.Info().Str("camera", raw.Name).Msg("seed: inserted camera")
		}
		seeded++
	}

	// Delete cameras that were removed from the YAML.
	yamlNames := map[string]bool{}
	for _, raw := range doc.Cameras {
		yamlNames[raw.Name] = true
	}
	for name, cam := range existingByName {
		if !yamlNames[name] {
			if _, err := repo.Delete(ctx, cam.ID); err != nil {
				return false, err
			}
			log.Info().Str("camera", name).Msg("seed: removed camera not in YAML")
		}
	}

	if err := tx.Commit(); err != nil {
		return false, err
	}
	log.Info().Int("cameras", seeded).Msg("seed complete")
	return true, nil
}
